### PR TITLE
Added service message json codec classes

### PR DESCRIPTION
**Pull request by ronenhamias**
1. Create separate service json transport for service to service communication
2. Create api gateway support http and socketio 
3. integrate gateway to services to allow dispatching of services via http and socket.io

### CODE REVIEW

File: gateway/src/main/java/io/scalecube/gateway/http/HttpCodecUtil.java [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+
+public final class HttpCodecUtil {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(HttpCodecUtil.class);
+
+  public static final String HEADER_STREAM_ID = "X-Stream-ID";
+
+  private static final int INTERNAL_SERVER_ERROR_CODE = 500;
+
+  private HttpCodecUtil() {
```
Comment (antonkharenko) [inline gateway/src/main/java/io/scalecube/gateway/http/HttpCodecUtil.java]: This class seems to be useless

File: ipc/src/main/java/io/scalecube/ipc/netty/NettyBootstrapFactory.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import io.netty.channel.nio.NioEventLoopGroup;
+import io.netty.channel.socket.ServerSocketChannel;
+import io.netty.channel.socket.SocketChannel;
+import io.netty.channel.socket.nio.NioServerSocketChannel;
+import io.netty.channel.socket.nio.NioSocketChannel;
+import io.netty.util.concurrent.DefaultThreadFactory;
+
+import java.util.concurrent.ThreadFactory;
+
+public final class NettyBootstrapFactory {
```
Comment (antonkharenko) [inline ipc/src/main/java/io/scalecube/ipc/netty/NettyBootstrapFactory.java]: Consider an opportunity to unify it with BoostrapFactory at transport module (not this PR)

File: ipc/src/main/java/io/scalecube/ipc/netty/ServiceChannelInitializer.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+import io.netty.channel.ChannelHandler.Sharable;
+import io.netty.channel.ChannelInitializer;
+import io.netty.channel.ChannelPipeline;
+import io.netty.handler.codec.LengthFieldBasedFrameDecoder;
+import io.netty.handler.codec.LengthFieldPrepender;
+
+@Sharable
+public final class ServiceChannelInitializer extends ChannelInitializer {
+
+  private static final int LENGTH_FIELD_LENGTH = 2;
```
Comment (antonkharenko) [inline ipc/src/main/java/io/scalecube/ipc/netty/ServiceChannelInitializer.java]: Why 2 instead of 4 bytes?

Comment (artem-v) [reply ipc/src/main/java/io/scalecube/ipc/netty/ServiceChannelInitializer.java]: Ha ha. You noticed!  I will set it to 4.

File: ipc/src/main/java/io/scalecube/ipc/netty/NettyBootstrapFactory.java (hunk 2 above)
Comment (artem-v) [reply ipc/src/main/java/io/scalecube/ipc/netty/NettyBootstrapFactory.java]: I was thinking to have separate one, not to depend on `scalecube-transport` in new modules: `gateway` and `ipc`.

Comment (antonkharenko) [reply ipc/src/main/java/io/scalecube/ipc/netty/NettyBootstrapFactory.java]: I was not speaking about adding this dependency, but more like to move this stuff into `utils` or some other common module.

File: gateway/src/main/java/io/scalecube/gateway/GatewayRunner.java [hunk 4]
```diff
@@ -0,0 +1,21 @@
+package io.scalecube.gateway;
+
+import io.scalecube.gateway.http.GatewayHttpServer;
+import io.scalecube.gateway.socketio.GatewaySocketIOServer;
+import io.scalecube.ipc.ServerStream;
+import io.scalecube.ipc.netty.NettyBootstrapFactory;
+
+public final class GatewayRunner {
```
Comment (ronenhamias) [inline gateway/src/main/java/io/scalecube/gateway/GatewayRunner.java]: Move this class to examples

File: ipc/src/main/java/io/scalecube/ipc/IpcEvent.java [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+    return new Builder(other);
+  }
+
+  //// Getters
+
+  public Topic getTopic() {
+    return topic;
+  }
+
+  public Optional<IpcMessage> getMessage() {
```
Comment (ronenhamias) [inline ipc/src/main/java/io/scalecube/ipc/IpcEvent.java]: rename to message()

File: ipc/src/main/java/io/scalecube/ipc/IpcEvent.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+    return message;
+  }
+
+  public IpcMessage justGetMessage() {
```
Comment (ronenhamias) [inline ipc/src/main/java/io/scalecube/ipc/IpcEvent.java]: remove this getter

File: gateway/src/main/java/io/scalecube/gateway/GatewayEchoRunner.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
+package io.scalecube.gateway;
+
+import io.scalecube.gateway.http.GatewayHttpServer;
+import io.scalecube.gateway.socketio.GatewaySocketIOServer;
+import io.scalecube.ipc.MessageStream;
+import io.scalecube.ipc.ServerStream;
+
+public final class GatewayEchoRunner {
+
+  public static void main(String[] args) throws Exception {
```
Comment (ronenhamias) [inline gateway/src/main/java/io/scalecube/gateway/GatewayEchoRunner.java]: move to examples

File: gateway/src/main/java/io/scalecube/gateway/GatewayRunner.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import io.scalecube.gateway.http.GatewayHttpServer;
+import io.scalecube.gateway.socketio.GatewaySocketIOServer;
+import io.scalecube.ipc.ClientStream;
+import io.scalecube.ipc.MessageStream;
+import io.scalecube.ipc.ServerStream;
+import io.scalecube.transport.Address;
+
+public final class GatewayRunner {
+
+  public static void main(String[] args) throws Exception {
```
Comment (ronenhamias) [inline gateway/src/main/java/io/scalecube/gateway/GatewayRunner.java]: move to examples

**Pull request closed without merging**